lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -1987,39 +1977,29 @@ impl<'a> FuncGen<'a> {
         local_count: u32,
         local_type: WpType,
     ) -> Result<(), CodegenError> {
-        self.local_count =
-            self.local_count
-                .checked_add(local_count)
-                .ok_or_else(|| CodegenError {
-                    message: "too many locals and arguments".into(),
-                })?;
-        self.local_types.push((self.local_count - 1, local_type));
+        self.local_types
+            .push(local_count, local_type)
+            .map_err(|e| CodegenError {
+                message: format!("could not process a local: {}", e),
+            });
+        // We will want to add +1 to this number in `local_count` later.
+        if self.local_types.max_index() == Some(u32::max_value()) {
```

Comment:
PartialSumMap is capable of storing `T::max_value() + 1` values since it deals with indices and not element counts. I did ponder making an API for a count too, but it definitely gets weird (what the return value should be for `local_types.local_count()`?) when the API for the same data structure tries to deal with both indices and element counts.

I ended up punting on this question as seen here and I think it in the end resulted in a properly scoped data structure, even if the calling code needs to deal with hacks like these.

Reply:
Makes sense!